cmd/optimize/events.go

```diff
@@ -0,0 +1,588 @@
+  fsoc optimize events --since -7d until 2023-07-31
+  fsoc optimize events --events="experiment_deployment_started,experiment_deployment_completed"
+  fsoc optimize events --optimizer-id namespace-name-00000000-0000-0000-0000-000000000000 --count 5
+  fsoc optimize events --namespace some-namespace --cluster-id 00000000-0000-0000-0000-000000000000
+  fsoc optimize events --workload-name some-workload`,
+		RunE:             listEvents(&flags),
+		TraverseChildren: true,
+		Annotations: map[string]string{
+			output.TableFieldsAnnotation:  "OptimizerId: .EventAttributes[\"optimize.optimization.optimizer_id\"], EventType: .EventAttributes[\"appd.event.type\"], Timestamp: .Timestamp",
+			output.DetailFieldsAnnotation: "OptimizerId: .EventAttributes[\"optimize.optimization.optimizer_id\"], EventType: .EventAttributes[\"appd.event.type\"], Timestamp: .Timestamp, Attributes: .EventAttributes",
+		},
+	}
+
+	command.Flags().StringVarP(&flags.clusterId, "cluster-id", "c", "", "Retrieve events constrained to a specific cluster by its ID")
+	command.Flags().StringVarP(&flags.namespace, "namespace", "n", "", "Retrieve events constrained to a specific namespace by its name")
+	command.Flags().StringVarP(&flags.workloadName, "workload-name", "w", "", "Retrieve events constrained to a specific workload by its name")
+	command.Flags().StringVarP(&flags.optimizerId, "optimizer-id", "i", "", "Retrieve events for a specific optimizer by its ID")
+	command.MarkFlagsMutuallyExclusive("optimizer-id", "cluster-id")
+	command.MarkFlagsMutuallyExclusive("optimizer-id", "namespace")
+	command.MarkFlagsMutuallyExclusive("optimizer-id", "workload-name")
+
+	command.Flags().BoolVarP(&flags.includeProgress, "include-progress", "p", false, "Include progress events in query and output")
+	command.Flags().StringSliceVarP(&flags.events, "events", "e", defaultEvents, "Customize the types of events to be retrieved")
+	command.MarkFlagsMutuallyExclusive("include-progress", "events")
+
+	command.Flags().StringVarP(&flags.since, "since", "s", "", "Retrieve events contained in the time interval starting at a relative or exact time. (default: -1h)")
+	command.Flags().StringVarP(&flags.until, "until", "u", "", "Retrieve events contained in the time interval ending at a relative or exact time. (default: now)")
+
+	command.Flags().IntVarP(&flags.count, "count", "", -1, "Limit the number of events retrieved to the specified count")
+
+	command.Flags().StringVarP(&flags.solutionName, "solution-name", "", "optimize", "Intended for developer usage, overrides the name of the solution defining the FMM types for reading")
+	if err := command.LocalFlags().MarkHidden("solution-name"); err != nil {
+		log.Warnf("Failed to set events solution-name flag hidden: %v", err)
+	}
+
+	return command
+}
+
+type eventsTemplateValues struct {
+	Since  string
+	Until  string
+	Events string
+	Filter string
+	Limits string
+}
+
+var eventsTemplate = template.Must(template.New("").Parse(`
```

Comment:
AFAIK results are not guaranteed to be ordered, you can add this to force an order (important when you also have a limit):
```
ORDER
   events.desc() 
```

Reply:
I ended up having to use `events.asc()` for the ordering so that the latest events would show up at the bottom of the list which I find to be more readable and also important for the `--follow` implementation.

However this creates a complication with the `--count` parameter in that it no longer gets the latest n events. Instead it gets the first n events to occur after the SINCE boundary. Ultimately I think I will need to propose an update to output.go to allow reversing the order of the table being printed but want to get events merged in as is for now.